### PR TITLE
Detect app launch timeout on Apple

### LINKED ISSUE #785: Apple runs time out while on --launch-timeout but result is TIMED_OUT
### Context
We have two timeouts in XHarness. First one, `--launch-timeouts`, starts when XHarness starts and is discarded when the application boots and test execution starts (we have TCP connection). Second, `--timeout`, runs from the beginning and breaks the whole run when is hit.

### Problem
We see `TIMED_OUT` results after `--launch-timeout` starts but this timeout should give us `PACKAGE_INSTALLATION_TIMEOUT` probably because we never start executing.

Or we have started executing the app and this timeout should not matter anymore. We can then have `TIMED_OUT` if the second timeout for the whole run is hit.

We shouldn't have `TIMED_OUT` after `--launch-timeout` is over though. 

Example logs ([full logs](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-2e2e0e3f8d2b451881/Microsoft.Extensions.Configuration.UserSecrets.Tests/1/console.520436bf.log?sv=2019-07-07&se=2021-12-20T20%3A38%3A00Z&sr=c&sp=rl&sig=9BAEK6uSnuR7e%2B0BcCz9WYrki0vFX8fpRtD%2B3qH9u8c%3D))
```
XHarness command issued: apple test --app /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app --output-directory /tmp/helix/working/A62F0846/w/A14608E7/uploads --target tvos-simulator --timeout 00:30:00 --xcode /Applications/Xcode115.app -v --launch-timeout 00:03:00
[12:38:15] info: Preparing run for tvos-simulator
[12:38:15] info: Getting app bundle information from '/tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app'
[12:38:15] dbug: 
[12:38:15] dbug: Running /usr/libexec/PlistBuddy -c "Print CFBundleName" /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app/Info.plist
[12:38:15] dbug: Process PlistBuddy exited with 0
[12:38:15] dbug: 
[12:38:15] dbug: Running /usr/libexec/PlistBuddy -c "Print CFBundleIdentifier" /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app/Info.plist
[12:38:15] dbug: Process PlistBuddy exited with 0
[12:38:15] dbug: 
[12:38:15] dbug: Running /usr/libexec/PlistBuddy -c "Print UIRequiredDeviceCapabilities" /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app/Info.plist
[12:38:15] dbug: Process PlistBuddy exited with 1
[12:38:15] dbug: Property UIRequiredDeviceCapabilities not present in Info.plist, assuming 32-bit is not supported
[12:38:15] info: Looking for available tvos-simulator simulators..
[12:38:15] dbug: Looking for available tvos-simulator simulators. Storing logs into list-tvos-simulator-20211130_123815.log
[12:38:32] info: Found simulator device 'Apple TV (tvOS 13.4) - created by XHarness'
[12:38:32] info: Uninstalling any previous instance of 'net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests' from 'Apple TV (tvOS 13.4) - created by XHarness'
[12:38:32] dbug: 
[12:38:32] dbug: Running /Applications/Xcode115.app/Contents/Developer/usr/bin/simctl uninstall F564F86E-0133-4B19-99A9-044549D844D8 net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests
[12:38:32] dbug: An error was encountered processing the command (domain=com.apple.CoreSimulator.SimError, code=165):
[12:38:32] dbug: Unable to lookup in current state: Shutdown
[12:38:32] dbug: Process simctl exited with 165
[12:38:32] dbug: Preemptive uninstallation of application failed
[12:38:32] info: Installing application 'Microsoft.Extensions.Configuration.UserSecrets.Tests' on 'Apple TV (tvOS 13.4) - created by XHarness'
[12:38:32] dbug: Installing '/tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app' to 'Apple TV (tvOS 13.4) - created by XHarness' (46.84 MB)
[12:38:32] dbug: 
[12:38:32] dbug: Running /private/tmp/helix/working/A62F0846/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.21576.2/tools/net6.0/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app --device=:v2:udid=F564F86E-0133-4B19-99A9-044549D844D8 --installsim /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app -v -v
[12:38:33] dbug: Using Xcode 11.5 found in /Applications/Xcode115.app
[12:38:33] dbug: Xamarin.Hosting: Xamarin.Hosting
[12:38:33] dbug: Xamarin.Hosting:     Version: efad4f0bae (main)
[12:38:33] dbug: Xamarin.Hosting:     Xcode: /Applications/Xcode115.app
[12:38:33] dbug: Xamarin.Hosting:     Xcode Version: 11.5
[12:38:33] dbug: Xamarin.Hosting:     Verbosity: 2
[12:38:34] dbug: Xamarin.Hosting: Booting Apple TV (tvOS 13.4) - created by XHarness...
[12:38:35] dbug: Xamarin.Hosting: Booted Apple TV (tvOS 13.4) - created by XHarness successfully.
[12:38:35] dbug: Xamarin.Hosting: Installing /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app with Bundle Identifier net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests on 'tvOS 13.4 (17L255) - Apple TV'...
[12:39:50] dbug: Xamarin.Hosting: Installed 'net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests' from /tmp/helix/working/A62F0846/w/A14608E7/e/Microsoft.Extensions.Configuration.UserSecrets.Tests.app
[12:39:51] dbug: Xamarin.Hosting: Can't process potentially bad load commands, because the app is signed.
[12:39:51] dbug: Xamarin.Hosting: The bundle id net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests was successfully installed.
[12:39:57] dbug: Xamarin.Hosting: Shutdown Apple TV (tvOS 13.4) - created by XHarness successfully.
[12:39:57] dbug: Process mlaunch exited with 0
[12:39:57] info: Application 'Microsoft.Extensions.Configuration.UserSecrets.Tests' was installed successfully on 'Apple TV (tvOS 13.4) - created by XHarness'
[12:39:57] info: Starting test run for net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests..
[12:39:57] dbug: Test log server listening on: 0.0.0.0:55477
[12:39:57] dbug: *** Executing 'Microsoft.Extensions.Configuration.UserSecrets.Tests' on tvos-simulator 'Apple TV (tvOS 13.4) - created by XHarness' ***
[12:39:57] dbug: System log for the 'Apple TV (tvOS 13.4) - created by XHarness' simulator is: /private/var/root/Library/Logs/CoreSimulator/F564F86E-0133-4B19-99A9-044549D844D8/system.log
[12:39:57] dbug: Starting test run
[12:39:57] dbug: 
[12:39:57] dbug: Running /private/tmp/helix/working/A62F0846/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.21576.2/tools/net6.0/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app -setenv=NUNIT_AUTOEXIT=true -setenv=NUNIT_HOSTPORT=55477 -setenv=NUNIT_ENABLE_XML_OUTPUT=true -setenv=NUNIT_XML_VERSION=xUnit -setenv=NUNIT_HOSTNAME=127.0.0.1 --device=:v2:udid=F564F86E-0133-4B19-99A9-044549D844D8 --stdout=/tmp/helix/working/A62F0846/w/A14608E7/uploads/net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests.log --stderr=/tmp/helix/working/A62F0846/w/A14608E7/uploads/net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests.log --launchsimbundleid=net.dot.Microsoft.Extensions.Configuration.UserSecrets.Tests -v -v -v -v -v
[12:39:58] dbug: Using Xcode 11.5 found in /Applications/Xcode115.app
[12:39:58] dbug: Xamarin.Hosting: Xamarin.Hosting
[12:39:58] dbug: Xamarin.Hosting:     Version: efad4f0bae (main)
[12:39:58] dbug: Xamarin.Hosting:     Xcode: /Applications/Xcode115.app
[12:39:58] dbug: Xamarin.Hosting:     Xcode Version: 11.5
[12:39:58] dbug: Xamarin.Hosting:     Verbosity: 5
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTFoundation.framework/Versions/A/DVTFoundation
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTiPhoneSimulatorRemoteClient.framework/DVTiPhoneSimulatorRemoteClient
[12:39:58] dbug: Xamarin.Hosting: Loaded /Library/Developer/PrivateFrameworks/CoreSimulator.framework/CoreSimulator
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTDeviceKitBase.framework/DTDeviceKitBase
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTKit.framework/DVTKit
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTDeviceKit.framework/DTDeviceKit
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTXConnectionServices.framework/DTXConnectionServices
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTSourceControl.framework/DVTSourceControl
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTServices.framework/DVTServices
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTPortal.framework/DVTPortal
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTDocumentation.framework/DVTDocumentation
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTAnalyticsClient.framework/DVTAnalyticsClient
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/ContentDeliveryServices.framework/ContentDeliveryServices
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTAnalytics.framework/DVTAnalytics
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/SourceKit.framework/SourceKit
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/Frameworks/IDEFoundation.framework/IDEFoundation
[12:39:58] dbug: Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/IDEProducts.framework/IDEProducts
[12:39:58] dbug: Xamarin.Hosting: Simulator watchdogs are already disabled for 'tvOS 13.4 (17L255) - Apple TV'.
[12:39:59] dbug: Xamarin.Hosting: Launching simulator application 'com.apple.iphonesimulator'
[12:41:15] dbug: Killing process 2045 as it was cancelled
[12:41:15] dbug: Test execution timed out after 1.3038806166666668 minute(s).
[12:41:15] dbug: Process mlaunch exited with 137
[12:41:15] dbug: Test run failed
[12:41:15] dbug: Could not find pid in mtouch output.
[12:41:16] dbug: Test run never launched
[12:41:16] dbug: No crash reports, waiting 5 seconds for the crash report service...
[12:41:22] warn: Application test run timed out
```

### PR DESCRIPTION
**Pull request by premun**
The shared Xamarin code allowed for a short window in-between the app being installed and launched but not having connected over TCP to XHarness' listener.
We add a more granular code that will describe this case so that we don't need to return `TIMED_OUT` when `--launch-timeout` occurs since `TIMED_OUT` means that tests themselves are running too long. This new launch timeout is then categorized as infra failure rather than user's fault.

Resolves #785

### CODE REVIEW

File: src/Microsoft.DotNet.XHarness.Common/CLI/ExitCode.cs [hunk 1]
```diff
@@ -105,4 +105,10 @@ public enum ExitCode
     /// Hardware device is in some corrupted state, or just locked screen
     /// </summary>
     DEVICE_FAILURE = 89,
+
+    /// <summary>
+    /// This timeout occurs when the Apple app has been launched but hasn't
+    /// connected yetover TCP and --launch-timeout expires
+    /// </summary>
+    APP_LAUNCH_TIMEOUT = 90,
```
Comment (tkapin) [inline src/Microsoft.DotNet.XHarness.Common/CLI/ExitCode.cs:113]: Is this specific to Apple only? If so, shouldn't this be reflected in the enum name, or you want to have a generic category for this kind of faults even though we observe this just on Apple?

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs [hunk 2]
```diff
@@ -13,29 +13,30 @@ public enum TestExecutingResult
...
+    StateMask = NotStarted | InProgress | Waiting | Finished,
 
     // In progress state
-    Building = 0x10 + InProgress,
-    BuildQueued = 0x10 + InProgress + Waiting,
-    Built = 0x20 + InProgress,
-    Running = 0x40 + InProgress,
-    RunQueued = 0x40 + InProgress + Waiting,
-    InProgressMask = 0x10 + 0x20 + 0x40,
+    Building = 0x10 | InProgress,
```
Comment (tkapin) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: Why don't we have the basic flags (0x10, 0x20, 0x40) also denoted here? This makes the semantics of the compound flags really hard to understand. In other words, every compound flag (defined via the or'ing) should be expressed in terms of other named flags, not a "magic" constants.

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs [hunk 3]
```diff
@@ -617,7 +625,15 @@ public async Task<(TestExecutingResult ExecutingResult, string? ResultMessage)>
 
         if (_timedout)
         {
-            result.ExecutingResult = TestExecutingResult.TimedOut;
+            // Did the test execution start?
+            if (_listener.ConnectedTask.IsCompleted && _listener.ConnectedTask.Result)
```
Comment (tkapin) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs:629]: This is the same condition as above, wouldn't it be more readable to abstract this to some function with a descriptive name?

File: tests/Microsoft.DotNet.XHarness.iOS.Shared.Tests/TestReporterTests.cs [hunk 4]
```diff
@@ -368,4 +388,43 @@ public async Task ParseResultTimeoutTestsTest()
...
+        // false means we never received a connection which would flip it to true
+        testReporter.LaunchCallback(Task.FromResult(false));
+
+        var (result, resultMessage) = await testReporter.ParseResult();
+        Assert.Equal(TestExecutingResult.LaunchTimedOut, result);
+
+        // verify that we do not try to kill a process that never got started
+        _processManager.Verify(p => p.KillTreeAsync(It.IsAny<int>(), It.IsAny<ILog>(), true), Times.Never);
+        _crashReporter.Verify(c => c.EndCaptureAsync(TimeSpan.FromSeconds(5)), Times.Once);
+    }
```
Comment (tkapin) [inline tests/Microsoft.DotNet.XHarness.iOS.Shared.Tests/TestReporterTests.cs:429]: Nice, thumbs up for tests!

**Review by tkapin — APPROVED** [review]

File: src/Microsoft.DotNet.XHarness.Common/CLI/ExitCode.cs (hunk 1 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.Common/CLI/ExitCode.cs:113]: Yeah, it is possible we will re-use this in Android in future. We only agreed to have the description reflect the scenarios each code is used in.

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs (hunk 2 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: I am not sure I follow what you propose here?
The way I understand it as that `0x10` means `Building` which is also `InProgress`.

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs (hunk 3 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs:629]: Good idea, will do

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs [hunk 5]
```diff
@@ -296,8 +296,16 @@ public void LaunchCallback(Task<bool> launchResult)
...
-            _mainLog.WriteLine("Test execution timed out after {0} minute(s).", _timeoutWatch.Elapsed.TotalMinutes);
             _timedout = true;
+
+            if (_listener.ConnectedTask.IsCompleted && _listener.ConnectedTask.Result)
+            {
+                _mainLog.WriteLine($"Test execution timed out after {_timeoutWatch.Elapsed.TotalMinutes:0.##} minutes");
+            }
+            else
+            {
+                _mainLog.WriteLine($"Test failed to start in {_timeoutWatch.Elapsed.TotalMinutes:0.##} minutes");
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs:307]: I guess this was the Xamarin style. I agree, will do

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs (hunk 2 above)
Comment (tkapin) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: I got that, although technically, Building is 0x11 which is composed of <something> and InProgress. What's this <something>? What I'm after here is avoiding using the non-defined constants for clarity. Also the "Built = 0x20 | InProgress" seems weird as I'd expect Built be also Finished, which is not the case here. To be concrete, I could think of something like:
```
Build = 0x10,
BuildInProgress = Build | InProgress,
BuildFinished = Build | Finished,
BuildWaiting = Build | Waiting,
BuildQueued = Build | Waiting | InProgress
...
```
Similarly, I'd see e.g., "BuildSuceeded" defined not as magic `0x10000` and Succeeded, but as `Build | Succeeded`.
In other words, we seem to have several dimensions (States, Stages {Build, Run, perhaps Launch} and Results) projected to one, but the dimensions seem to not be aligned / defined right.

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: Gotcha. I agree that would be better, however this is coming from Xamarin and I don't think I can change this

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: I can maybe refactor it a little though

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: I tried to do what you propose and I think your example makes sense, however it is not applicable to the enum we have in the code.

You could sensibly apply your strategy to a couple maybe three of the values there but even then it would lead to an explosion of values of the enum which is probably counter-productive?

I also cannot reasonably apply it without refactoring the usage, example of some of the problems:
- You have `Succeeded = 0x100 | Finished` so you could factor out `Success = 0x100` and make `Succeeded = Success | Finished`
- However, `Success` would ideally belong in between these:
    ```
    NotStarted = 0,
    InProgress = 0x1,
    Finished = 0x2,
    Waiting = 0x4,
    ```
- But for that I would already have to change the value.
- Even if I did that, the original enum clashes later where this wouldn't make sense because for example `BuildSucceeded = 0x10000 | Succeeded` would ideally be something like `Build | Succeeded` but since we already have a new value for that `0x10000`, we cannot do that atm.

I think the whole enum would need to be built from ground up again to make proper sense but even then, I don't think the way it is used cares about something like "x is a Build" so we would only introduce an explosion of values in the enum and it might become quite hard to use.

Comment (tkapin) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestExecutingResult.cs:19]: Thanks for such detailed answer. Since the enum is already existing and being used, I agree with not refactoring it. 

I think you understood well what I'm after here, so just whenever we are building complex flags, please be mindful about the dimensions we are trying so fit (such as type of operation, state, result, etc.) as it's easy to clutter this as above.

**Review by rolfbjarne — APPROVED** [review]
